File: leetcode/LC/graph/dfs/0417_pacific_atlantic_water_flow.py
"""
O(m*n) | O(m*n) [ for visit cells]

for first row and first col do dfs for pacific ocean
for last row and last col do dfs for atlantic ocean
in dfs check if the value at row, col can reach previous value, so, previous value is passed as parameter.
current value should not be greater than previous value, if it is then return else do dfs on adjacent value of that row, col
"""
from typing import List

class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        
        row = len(heights)
        col = len(heights[0])
        p = set()
        a = set()
        
        def dfs(r, c, visit, height):
            if (r,c) in visit or r < 0 or r >= row or c < 0 or c >= col or height > heights[r][c]:
                return
            visit.add((r,c))
            
            directions = [(r+1, c), (r-1, c), (r, c+1), (r, c-1)]
            for d in directions:
                dfs(d[0], d[1], visit, heights[r][c])
        
        for c in range(col):
            dfs(0, c, p, heights[0][c])
            dfs(row-1, c, a, heights[row-1][c])
            
        
        for r in range(row):
            dfs(r,0, p, heights[r][0])
            dfs(r,col-1, a, heights[r][col-1])
        
        # do intersection of pacific and atlantic set
        ans = [v for v in p if v in a]
        return ans

File: leetcode/LC1/graph/0417_pacific_atlantic_water_flow.cpp
#include <climits>
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <queue>
#include <unordered_map>
using namespace std;


class Solution {
public:
 
    int maxRow;
    int maxCol;
    void dfs(int r, int c,  vector<vector<bool>>& v, int& height, vector<vector<int>>& heights){
        if(r < 0 or r > maxRow or c < 0 or c > maxCol or heights[r][c] < height or v[r][c])
           return;
        v[r][c] = true;
        vector<vector<int>> directions = {{r+1, c}, {r-1, c}, {r, c+1}, {r, c-1}};
           for(auto dir: directions){
               dfs(dir[0], dir[1], v, heights[r][c], heights);
           }
    }
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        
        int r = heights.size();
        int c = heights[0].size();
        maxRow = r-1;
        maxCol = c-1;
        vector<vector<bool>> p(r, vector<bool>(c, false));
        vector<vector<bool>> a(r, vector<bool>(c, false));
        for(int i = 0; i< heights[0].size(); i++){
            dfs(0, i, p, heights[0][i], heights);
            dfs(r-1, i, a, heights[r-1][i], heights);
        }
        for(int i = 0; i< heights.size(); i++){
            dfs(i, 0, p, heights[i][0], heights);
            dfs(i, c-1, a, heights[i][c-1], heights);
        }
        vector<vector<int>>ans;
        for(int i = 0; i< r; i++){
            for(int j = 0; j< c; j++){
                if(p[i][j] && a[i][j])
                    ans.push_back({i,j});
            }
        }
        return ans;        
    }
};
